### PR TITLE
Add support for vscode

### PR DESCRIPTION
**Pull request by jonnysoe**

### CODE REVIEW

File: .vscode/extensions.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+    // See https://go.microsoft.com/fwlink/?LinkId=827846 to learn about workspace recommendations.
+    // Extension identifier format: ${publisher}.${name}. Example: vscode.csharp
+
+    // List of extensions which should be recommended for users of this workspace.
+    "recommendations": [
+        "ms-vscode.cpptools-extension-pack",        // Comes with CMake for Debugging C++
```
Comment (GitMensch) [inline .vscode/extensions.json:7]: I personally won't recommend those proprietary extensions (only daohong-emilio.yash is free to use), which hinders me to pull this. Please drop all but yash (I wasn't even aware of it, thanks for sharing!) and consider adding cmake specific extension.

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
+        {
+            // https://code.visualstudio.com/docs/cpp/config-msvc
+            // https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools
+            "name": "(windbg) Launch",
+            "type": "cppvsdbg",
```
Comment (GitMensch) [inline .vscode/launch.json:11]: similar, this one is sadly not free; you may want to use clangd for intellisense and native-debug for debugging when you want to add those to the repo

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (GitMensch) [inline .vscode/settings.json:1]: It seems those settings would best be user-scope, not specific to this workspace.
Please share your reasoning if your disagree.

File: README.md [hunk 4]
```diff
@@ -25,10 +25,37 @@ The release page includes the full Changelog but you may also see the [changelog
 
 ## Build requirements
 * Visual Studio 2017 or newer
-* optional: CMake (when building with CMake)
+* optional: CMake (when building with CMake), Ninja
 
 ## HowTo
 You may use win_flex and win_bison directly on the command line or [use them via CustomBuildRules in VisualStudio](custom_build_rules/README.md).
+### Visual Studio Code
```
Comment (GitMensch) [inline README.md:32]: Following the current README's and repo structure this file and images should be moved out to a separate vscode folder, linked here as done for the CustomBuildRules.  
Please also consider using "vscode" instead, because there are more binary packages then the one distributed by Microsoft.

File: README.md [hunk 5]
```diff
@@ -8,7 +8,7 @@ Both win_flex and win_bison are based on upstream sources but depend on system l
 * 2.5.x versions include GNU Bison version 3.x.x
 
 ## License
-Flex uses a [BSD license](flex/src/COPYING), GNU Bison is [licensed under the GNU General Public License (GPLv3+)](bison/src/COPYING).  
+Flex uses a [BSD license](flex/src/COPYING), GNU Bison is [licensed under the GNU General Public License (GPLv3+)](bison/src/COPYING).
```
Comment (GitMensch) [inline README.md:11]: .md files should not be whitespace-trimmed at the end. Those two trailing spaces lead to a line break, otherwise it would be a block text, please revert this part.

File: README.md [hunk 6]
```diff
@@ -25,10 +25,37 @@ The release page includes the full Changelog but you may also see the [changelog
 
 ## Build requirements
 * Visual Studio 2017 or newer
```
Comment (GitMensch) [inline README.md:27]: We should change that to say "... build tools" (which is a separate package, for example pre-installed in several CI environments) or specify builds; this is not related to this PR but _could_ be done together.

File: README.md [hunk 7]
```diff
@@ -25,10 +25,37 @@ The release page includes the full Changelog but you may also see the [changelog
...
+5. Both `win_flex` and `win_bison` requires arguments for the target input/output source files, just add the arguments as how `win_flex` or `win_bison` was called from the command line where each space-separated argument is an element in [.vscode/launch.json](.vscode/launch.json)'s `args` array separated by commas, here is a sameple for `win_flex` target:
+```
+            "args": [
+                "--header-file=sample.tab.h",
+                "-osample.tab.c",
+                "sample.l"
+            ],
+```
+
+6. Setup the IDE for inspection [debugging](https://code.visualstudio.com/docs/editor/debugging), eg. breakpoints, logpoints.
```
Comment (GitMensch) [inline README.md:56]: it likely is more reasonable to not use "IDE" here; not sure if "editor" as used in the vscode docs is the best option, but it would be better.

Wording/spacing/punctuation:  "Set up" and "e.g. breakpoints or logpoints."

**Review by GitMensch — CHANGES_REQUESTED** [review]
Thank you for taking the time to document this. Please move the docs to a separate "vscode" folder, following the current docs structure.

Furthermore I do have concerns related to the proprietary extensions used, and I won't pull it myself as long as those are used. Lexxmark may see this different.

File: README.md (hunk 5 above)
Comment (jonnysoe) [reply README.md:11]: Its trimmed by trimTrailingWhitespace, I'll reinsert them and add `md` files as exception

File: .vscode/extensions.json (hunk 1 above)
Comment (jonnysoe) [reply .vscode/extensions.json:7]: I'll limit this to `ms-vscode.cmake-tools`, this alone is minimal and enough for normal use case

File: README.md (hunk 6 above)
Comment (jonnysoe) [reply README.md:27]: The pre-installed build tools I see in AppVeyor are all VS...
What does "... build tools" mean? Does that mean "Visual Studio 2017 or newer `build tools`"?
Or you you wanted something different like the following?
```
C++ build tools (MS BuildTools, Visual Studio 2017 or newer)
```

File: README.md (hunk 4 above)
Comment (jonnysoe) [reply README.md:32]: I'll try to mirror that directory...

File: README.md (hunk 7 above)
Comment (jonnysoe) [reply README.md:56]: How about calling it "codebase"? It'll be vague but indicative

File: README.md (hunk 6 above)
Comment (GitMensch) [reply README.md:27]: I mean - you don't need the full Visual Studio installed, the [Build Tools for C++](https://visualstudio.microsoft.com/de/visual-cpp-build-tools/) (which are included in Visual Studio) are enough.

File: README.md (hunk 7 above)
Comment (GitMensch) [reply README.md:56]: I'm fine with that.

File: .vscode/launch.json (hunk 2 above)
Comment (jonnysoe) [reply .vscode/launch.json:11]: This is the launcher for debugger, it has nothing to do with `clangd` as it is the language server for code navigation, and I'll leave `clangd` out since its interfering with IntelliSense, may cause problems for others too, IntelliSense will be removed from the workspace recommended `extensions.json`.

The debugger may be problematic, I'm not entirely sure about the `free` licensing you're referring to or what `native-debug` is, however this is the only debugger for MSVC which is what we're using, its not entirely `WinDbg` per se, but the same free proprietary tools Windows SDK bundles - Debugging TOols for Windows, WinDbg, CDB, KD, NTSD...

On a side note, did you mean open source? Please clarify

File: .vscode/settings.json (hunk 3 above)
Comment (jonnysoe) [reply .vscode/settings.json:1]: Workspace is the appropriate place for codebase consistency settings, unless that is something we don't want...

First is editor settings, most users have different settings which will make the codebase inconsistent, the proper way to have consistent code is clang-format, but that will format the whole code instead being just limited to the addtional lines introduced by developers; and will probably make it more difficult to sync with upstream.
Trailing whitespace is undesirable, for consistency and nothing much to explain further

Ruler indicator is just a mindful/useful indicator, I see GNU have an unwritten rule of 80 character line limit, but some codes go up to 120 chars in flex/bison, I'll say this is optional, I can remove this if you don't like to see these lines

The git operations are minimal and good to have for new developers, it helps to ensure git pull in vscode don't branch and merge, prune is optional as it helps to cleanup deleted branches in origin just that the code base is not super active at the moment

CMake preferred generator won't mess up the system as it only uses Ninja if its installed, although the codebase is small, it approximately halves the compilation time for debugging

File: README.md (hunk 6 above)
Comment (jonnysoe) [reply README.md:27]: Alright, its the latter, I'll update it :ok_hand:

File: .vscode/launch.json (hunk 2 above)
Comment (jonnysoe) [reply .vscode/launch.json:11]: @GitMensch ,
is your "free" referring to free-as-in-speech (libre)? [WinDbg](https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools) is part of Windows SDK which is free-as-in-beer (gratis), but its the only debugger for MSVC, its just configuration and we are not even redistributing this to users or developers

Comment (GitMensch) [reply .vscode/launch.json:11]: This is about the extension necessary - the cppvsdbg debug type is only part of a Microsoft extension that can only be used with Visual Studio Code, so "free beer if you sit only on a specific chair with no single adjustment, that comes pre-configured to tell the maker whenever you sit on it" (and the beer likely doing the same); and with the adjusted recommendations we don't have that available any more, I think. Am I wrong?

Note: If I remember correctly you can use LLDB to debug msvc generated modules, too - for this setup there are some free extensions (like "free to use for anything, anywhere, including adjusting the extension and re-distribute it"- "libre"), too.

Comment (jonnysoe) [reply .vscode/launch.json:11]: Nope, its a known issue which is why its never published in the vscode documentation, it may be possible in the CLI/TUI but LLDB cannot be used to debug MSVC-generated symbols in vscode as it [lacks LLDB-MI](https://github.com/microsoft/vscode-cpptools/issues/6617) for non-Apple platforms at the moment, the closest  LLDB for vscode is the [CodeLLDB](https://github.com/vadimcn/codelldb) extension which is not compatible with MSVC-based symbols; not even [MSVC-based Clang](https://github.com/llvm/llvm-project/releases/latest) programs, MSVC-based Clang programs will still need to use WinDbg since its PDB symbols are MSVC-based...
Getting [LLDB-MI](https://github.com/lldb-tools/lldb-mi) to work on Windows will require quite some effort from the community, which is inactive at the moment...

Yes, the recommended beer will be gone from the recommended extensions, I think having this debug configuration is not as evil as telling devs "the free beer is required or recommended", for now the codebase is only compatible with this beer (MSVC), if you need to debug, this is the only available beer (which we did not bundle) at the moment :speak_no_evil:

P.S. I can try to get the codebase compatible with [WinLibs](https://winlibs.com/)/UCRT after this, just not within this PR as it requires quite some changes

Comment (GitMensch) [reply .vscode/launch.json:11]: Note: Just tested, debugging x64 cl.exe generated programs (pdb) with the codelldb extension works just fine - as well as debugging msys2 mingw64 generated executables.

So, if you want to add a debug configuration what do you think about configuring that and add it to the suggested extensions?

Looking on the internet msys2 mingw-w64-x86_64-lldb-mi (if you don't have llvm already setup then the dependencies lead to a 916MB install!) seems to work fine for both GNU and PDB, too - using any vscode extension that supports this, like Native Debug.

Comment (jonnysoe) [reply .vscode/launch.json:11]: Can you check out strings and vectors?

Comment (GitMensch) [reply .vscode/launch.json:11]: C strings did work, as I'm currently working on C workspaces only I have no setup for `std::` - but if you get down to test them please drop a note.

Comment (jonnysoe) [reply .vscode/launch.json:11]: C strings will definitely work, STLs don't, since upstream are GNU guys and they don't like C++ or STLs, I can accept codeLLDB without being able to debug `std::string` and `std::vector` in this workspace, I'll default this to codeLLDB

Comment (GitMensch) [reply .vscode/launch.json:11]: LLDB folks are GNU guys?
GNU guys don't like C++or STLs (just have a look what some of the GNU software is written in...)?

Note: for GDB you commonly want pretty-print handlers for C++ and STL types, maybe this is similar with LLDB?

Comment (GitMensch) [reply .vscode/launch.json:11]: > The debugger may be problematic, I'm not entirely sure about the `free` licensing you're referring to

the [microsoft extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools) is not free, as you can [see in its license](https://marketplace.visualstudio.com/items/ms-vscode.cpptools/license)

> You may install and use any number of copies of the software only with Microsoft [...] Visual Studio Code

--> this explicit disallows use with any other built of vscode repo and the extension also checks for the Microsoft identifiers in product.json, so disables itself when it "believes" it is not run on the "single blessed binary product"... and they even have a wiki page explaining that: https://github.com/microsoft/vscode-cpptools/wiki/Microsoft-Native-Debugger-licensing-and-Microsoft-Visual-Studio-Code

> or what `native-debug` is

It is one of the extensions that can work with lldb-mi, so _if_ you have a binary lldb-mi, then you can use it.

> however this is the only debugger for MSVC which is what we're using,

No, as noted you can use the lldb extension which distributes windows binaries, too (but not lldb-mi) to debug MSVC generated binaries as well as any debug extension that supports lldb-mi.

My point was _exactly_ that I don't this repo to suggest people to use the non-free extension (that will only work on "MS' blessed binaries, so for example not any of my vscode installations)and let all others have a note in their problems pane that the debug type is unsupported.

Comment (jonnysoe) [reply .vscode/launch.json:11]: Thank you so much for going above and beyond to clarify, now I understood what you meant, it is free if user/dev is on MS tool suites, not free for any other variants or forks of `Code - OSS`
TL;DR  the rule for this free beer is MS binaries

Comment (GitMensch) [reply .vscode/launch.json:11]: > TL;DR the rule for this free beer is MS binaries

yes, along with binaries where the code is explicit not released - and both have an explicit opt-out of telemetry (which you can set, but as the sources are closed you can only check what goes out by firewall [blocking some of that works, others broke the extension, but that's a too unrelated thing to this PR and I also have no recent data about that, so "ignore that part of my rant, please - just keep in mind 'nobody but MS can actually check what the software does'"]).

**Comment by jonnysoe** [comment]
@GitMensch , my apologies for leaving this hanging, I totally lost track of this ever since I left my last job...
I'll check that branch out to rebase it, gimme some time to follow through the conversations and new commits

**Comment by jonnysoe** [comment]
@GitMensch ,
I've read through some of the comments by you guys, ignoring .vscode is essentially defeating the purpose of this PR, it was supposed to introduce convenience to ease contributors for an easy start, the extensions.json doesn't force anything upon the end user either - it simply floats the recommended extensions for this code base to the top in the extensions tab, I understand if contributors accidentally merged unnecessary changes to the launch.json and I have no good way to prevent it - either add a generator script without committing the launch.json file or a git hook to prevent changes to it (thats how vscodium code base [prevents](https://github.com/microsoft/vscode/issues/31168) contributors from adding anything to extensionsGallery.
I'm leaning towards a generator (such as copy during first build); the git hook method is less reliable to enforce because it requires additional git commands to be functional - which we can also add a script during build but its easy to bypass when the contributor don't build at all. Let me know whats your preference.

File: .vscode/settings.json (hunk 3 above)
Comment (jonnysoe) [reply .vscode/settings.json:1]: @Febbe ,
> accidentally push private & personal settings / metadata


Personal settings stay within their user directory `settings.json` by default when changes are made from the GUI, so if the contributor manually does questionable changes for the code base, then catching questionable changes or they're justifiable is what reviewers are here for


> It doesn't necessarily format the whole base


I don't recall `git clang-format` limits changes to the modified lines back in llvm 15

**Comment by GitMensch** [comment]
@jonnysoe Sorry, I didn't understand your question.

Please do the doc adjustments - I think the point was to move some things around and to resolve the conflicts - then restate your question.

... and possibly make it more explicit what this PR is about:
* adjusting winflexbison itself from a checkout / source snapshot in vscode
* documenting how to use that in vscode with a project that uses bison/flex

... or both? If that is the case then the second part should be split into a separate PR which we likely can merge easily and then also have a smaller scope here.

... seems I'm confused.

**Comment by jonnysoe** [comment]
> the rest like extensions to be used in this repo will be ignored already

I understand the doc changes requested - move the images around, I was referring to this statement that doesn't make sense to me, the point is to add support for vscode (basically the `extensions.json` and `launch.json` introduced), the docs are basically guide to end users on how to use vscode to debug bison/flex.

Which is this
> documenting how to use that in vscode with a project that uses bison/flex


I don't understand this following statement
> adjusting winflexbison itself from a checkout / source snapshot in vscode

File: .vscode/settings.json (hunk 3 above)
Comment (Febbe) [reply .vscode/settings.json:1]: I’ve been using `git clang-format` for much longer, and its sole purpose is to only operate on the modified lines.
There used to be a run-clang-format.py script that formatted everything.
Additionally, you can configure clangd in VS Code as the formatter for C/C++, and restrict it to formatting only modified lines. That has the same effect as git-clang-format as clangd linked clang-format into its server.

File: .vscode/launch.json (hunk 2 above)
Comment (jonnysoe) [reply .vscode/launch.json]: Replaced VS Debugger with codelldb

File: .vscode/settings.json (hunk 3 above)
Comment (jonnysoe) [reply .vscode/settings.json:1]: Yes, I know of the extension being able to format by lines, I just couldn't get it to work when I use command line...

File: README.md (hunk 4 above)
Comment (jonnysoe) [reply README.md]: I've recaptured the screenshots and moved everything to the new `vscode` directory

File: README.md (hunk 6 above)
Comment (jonnysoe) [reply README.md]: I left this unchanged after rebase

File: README.md (hunk 7 above)
Comment (jonnysoe) [reply README.md]: Updated, see the new vscode/README.md

**Comment by jonnysoe** [comment]
@GitMensch ,
Please continue with the review. Thanks!

File: .vscode/extensions.json [hunk 8]
```diff
@@ -0,0 +1,11 @@
+{
+    // See https://go.microsoft.com/fwlink/?LinkId=827846 to learn about workspace recommendations.
+    // Extension identifier format: ${publisher}.${name}. Example: vscode.csharp
+
+    // List of extensions which should be recommended for users of this workspace.
+    "recommendations": [
+        "ms-vscode.cmake-tools",                    // CMake for Debugging C/C++
+        "vs-code-extensions.vscode-clangd",         // Optional intellisense-like features with LLVM/clangd
+        "daohong-emilio.yash",                      // For flex/bison source files
```
Comment (GitMensch) [inline .vscode/extensions.json]: This one is not available in open-vsx; I've searched and found a full-fledged new extension https://github.com/theodevelop/bison-flex-lang - should we recommend that?

File: .vscode/extensions.json [hunk 9]
```diff
@@ -0,0 +1,11 @@
+{
+    // See https://go.microsoft.com/fwlink/?LinkId=827846 to learn about workspace recommendations.
+    // Extension identifier format: ${publisher}.${name}. Example: vscode.csharp
+
+    // List of extensions which should be recommended for users of this workspace.
+    "recommendations": [
+        "ms-vscode.cmake-tools",                    // CMake for Debugging C/C++
+        "vs-code-extensions.vscode-clangd",         // Optional intellisense-like features with LLVM/clangd
```
Comment (GitMensch) [inline .vscode/extensions.json]: Why not the original `llvm-vs-code-extensions.vscode-clangd`?

File: .gitignore
Comment (GitMensch) [inline .gitignore:1]: Why do we want to ignore a folder we directly add in the same commit?

File: README.md [hunk 10]
```diff
@@ -35,7 +35,10 @@ The release page includes the full Changelog but you may also see the [changelog
 
 ## HowTo
 
-You may use win_flex and win_bison directly on the command line or [use them via CustomBuildRules in VisualStudio](custom_build_rules/README.md).
+You may:
+- use `win_flex` and `win_bison` directly on the command line
+- [use them via CustomBuildRules in VisualStudio](custom_build_rules/README.md).
+- use them via [CMake Tools](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cmake-tools) in [Visual Studio Code](vscode/README.md).
```
Comment (GitMensch) [inline README.md]: It seems to me that those would also be useful with the bison/flex extension noted above.

Comment (GitMensch) [inline README.md]: you may actually use them in vscode compatible environments (like an open build as VSCodium [used also as vscode in arch] or even Eclipse Thea) - I'd personally prefer replacing the product name of the MS binaries to just "vscode"...

File: vscode/README.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
+
+Alternatively, VS Code is available for debugging, here is a simple guide (based on the [MS documentation](https://code.visualstudio.com/docs/cpp/cmake-linux)):
```
Comment (GitMensch) [inline vscode/README.md]: A file starting with "Alternatively" seems bad.
If this file is only about how to debug winflexbison (which is fine and seems to be the case), we can just drop most of the first sentence and instead use a header like `# Debugging WinFlexBison in vscode`.

File: README.md (hunk 10 above)
Comment (GitMensch) [inline README.md]: The referenced file is not about how to use it in vscode, so we may drop the link altogether (and/or add another leading to `# How To Use WinFlexBison` and `# How To Compile/Debug WinFlexBison` where we reference the build scripts and the new vscode readme.

**Review by GitMensch — CHANGES_REQUESTED** [review]
Thanks for the iteration and the explanations up front.

I think adding .vscode itself does make sense and the documentation is also good - mostly wondering about the recommended extensions now...

File: .vscode/extensions.json (hunk 8 above)
Comment (jonnysoe) [reply .vscode/extensions.json]: This is fresh out of the oven for a week, just tested it feels a bit rough, very active development, looks good, so the name is `theodevelop.bison-flex-lang` on both repository, I'll update it
https://marketplace.visualstudio.com/items?itemName=theodevelop.bison-flex-lang
https://open-vsx.org/extension/theodevelop/bison-flex-lang

File: .vscode/extensions.json (hunk 9 above)
Comment (jonnysoe) [reply .vscode/extensions.json]: Not sure how did I screw up the copy, I'll update it

File: .gitignore
Comment (jonnysoe) [reply .gitignore:1]: As a deterrant, changes to this directory will have warnings, and developers will need to force add new files

File: README.md (hunk 10 above)
Comment (jonnysoe) [reply README.md]: I agree, there is some form of genericization, how does "vscode-like editor" sound?

Comment (jonnysoe) [reply README.md]: I'm not getting this statement, mind elaborating more?

File: vscode/README.md (hunk 11 above)
Comment (jonnysoe) [reply vscode/README.md]: Thanks, this is part of the vestige from original cut/paste, I've tried to update most of the things here, missed this 1

File: README.md (hunk 10 above)
Comment (jonnysoe) [reply README.md]: This falls under `How to Debug`, and implicitly `Compile`, how do you want to split this actually? Under the 2 headers you mentioned?

Comment (jonnysoe) [reply README.md]: @GitMensch 
Just rereading this, I believe you mean naming the flex/bison extension here, however the extension does not contribute to this `HowTo` (build/debug), this extension is purely convenience for the developer (goto definition, syntax highlighting, etc. basically things that makes code navigation a pleasant experience)
I don't think we should name it as required (hence its a recommended extension that appears in the extensions list), basically optional like telling ppl "you should use git for this code base", something that we don't need to spell out, it will work regardless.
Lemme know if you still wanna add it here, I can do that, its just a simple "and" statement.
